File: pegasusX/apps/factory-app-ios/FactoryApp/Utilities/GeocodeLocationSupport.swift
import Foundation

enum GeocodeLocationSupport {
    static func hasValidCoordinates(lat: Double, lng: Double) -> Bool {
        guard lat.isFinite, lng.isFinite else { return false }
        if lat == 0, lng == 0 { return false }
        return (-90 ... 90).contains(lat) && (-180 ... 180).contains(lng)
    }

    static func resolveLocationValue(_ value: AddressLocationValue) async -> AddressLocationValue? {
        let address = value.address.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !address.isEmpty else { return nil }
        if hasValidCoordinates(lat: value.lat, lng: value.lng) { return value }

        let predictions = (try? await GeocodeService.autocomplete(address)) ?? []
        if let top = predictions.first, !top.placeId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            if let byPlace = try? await GeocodeService.resolvePlace(top.placeId),
               hasValidCoordinates(lat: byPlace.lat, lng: byPlace.lng) {
                return AddressLocationValue(
                    address: byPlace.address.isEmpty ? address : byPlace.address,
                    lat: byPlace.lat,
                    lng: byPlace.lng,
                    placeId: byPlace.placeId
                )
            }
        }

        if let byAddress = try? await GeocodeService.forward(address: address),
           hasValidCoordinates(lat: byAddress.lat, lng: byAddress.lng) {
            return AddressLocationValue(
                address: byAddress.address.isEmpty ? address : byAddress.address,
                lat: byAddress.lat,
                lng: byAddress.lng,
                placeId: byAddress.placeId
            )
        }
        return nil
    }
}

File: pegasusX/apps/warehouse-app-android/app/src/main/java/com/pegasusx/warehouse/util/GeocodeLocationSupport.kt
package com.pegasusx.warehouse.util

import com.pegasusx.warehouse.data.remote.GeocodeApi
import com.pegasusx.warehouse.ui.components.AddressLocationValue

object GeocodeLocationSupport {
    fun hasValidCoordinates(lat: Double, lng: Double): Boolean {
        if (!lat.isFinite() || !lng.isFinite()) return false
        if (lat == 0.0 && lng == 0.0) return false
        return lat in -90.0..90.0 && lng in -180.0..180.0
    }

    suspend fun resolveLocationValue(
        geocodeApi: GeocodeApi,
        value: AddressLocationValue,
    ): AddressLocationValue? {
        val address = value.address.trim()
        if (address.isEmpty()) return null
        if (hasValidCoordinates(value.lat, value.lng)) return value

        val top = runCatching { geocodeApi.autocomplete(address).predictions.firstOrNull() }.getOrNull()
        if (!top?.placeId.isNullOrBlank()) {
            val byPlace = runCatching { geocodeApi.resolvePlace(top!!.placeId) }.getOrNull()
            if (byPlace != null && hasValidCoordinates(byPlace.lat, byPlace.lng)) {
                return AddressLocationValue(
                    address = byPlace.address.ifBlank { address },
                    lat = byPlace.lat,
                    lng = byPlace.lng,
                    placeId = byPlace.placeId,
                )
            }
        }

        val byAddress = runCatching { geocodeApi.forward(com.pegasusx.warehouse.data.model.ForwardGeocodeRequest(address = address)) }.getOrNull()
        if (byAddress != null && hasValidCoordinates(byAddress.lat, byAddress.lng)) {
            return AddressLocationValue(
                address = byAddress.address.ifBlank { address },
                lat = byAddress.lat,
                lng = byAddress.lng,
                placeId = byAddress.placeId,
            )
        }
        return null
    }
}
